[SPARK-4980] [MLlib] Add decay factors to streaming linear methods

rotationsymmetry opened the pull request:
This PR includes an implementation of decay factors in streaming linear and logistic regression. Unit tests are also included. 

The algorithm and design details are described in the document: https://docs.google.com/document/d/1UfKvuaaJVQCvh-wOLLYT8l7STQFjPxE7fitZyd0tqTo/edit?usp=sharing

Your comments and suggestions are highly appreciated. I will add more tests and ScalaDoc as suggested.

Thanks!

cc @freeman-lab @mengxr

rotationsymmetry committed:
Add decay to StreamingLinearAlgorithm through StreamingDecay trait.

Revise test "parameter accuracy" in StreamingLinearRegressionSuite to account for decay.

rotationsymmetry committed:
Fix fluent setter API

Split StreamingDecay into two traits.
Update StreamingLogisticRegressionWithSGD.
Update test suites.

rotationsymmetry committed:
Add unit tests.

Also make StreamingDecaySetter to be private[mllib].

rotationsymmetry committed:
minor fixes

AmplabJenkins commented:
Can one of the admins verify this patch?

rotationsymmetry committed:
fix Scala style

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,57 @@
+  def getDiscount(numNewDataPoints: Long): Double
+}
+
+private[mllib] trait StreamingDecaySetter[T <: StreamingDecaySetter[T]] extends Logging {
```
Why do we need F-bounded polymorphism here? Does the code not work when you replace `T` with `self.type`?

rotationsymmetry replied:
"Does the code not work when you replace T with self.type?"

I guess not? For example, 
``` scala
trait Setter { def set: this.type = this}
class Apple extends Setter
val a = new Apple()
a.set
```
The return type of `a.set` is `a.type`, not `Apple`. Do I answer your question?

"Why do we need F-bounded polymorphism here?"
I agree with you that this is not needed here. Originally I included this as an extra level of type checking. But since I have `self: T=>` in the next line, I don't think we need it any more. I will remove it in the next push to the PR.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,57 @@
+
+private[mllib] trait StreamingDecaySetter[T <: StreamingDecaySetter[T]] extends Logging {
+  self: T =>
+  var decayFactor: Double = 0
```
These public attributes need scaladocs describing them and their default values

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,57 @@
+    this
+  }
+
+  /** Set the half life and time unit ("batches" or "points") for forgetful algorithms. */
```
Provide full scaladoc with `@params` for all public APIs

rotationsymmetry replied:
Thanks! Will do.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,57 @@
+import org.apache.spark.Logging
+
+trait StreamingDecay {
+  def getDiscount(numNewDataPoints: Long): Double
```
Scaladocs for public APIs

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,57 @@
+private[mllib] trait StreamingDecaySetter[T <: StreamingDecaySetter[T]] extends Logging {
+  self: T =>
+  var decayFactor: Double = 0
+  var timeUnit: String = StreamingDecay.BATCHES
```
Can these be private?

rotationsymmetry replied:
I agree that the `StreamingDecay` trait can be `private[mllib]`. Will update this.

feynmanliang replied:
Oh didn't see that on L22, yeah I agree (thanks for catching :+1:)!

What I meant here is that since you provide getters and setters for `decayFactor` and `timeUnit` in the trait, perhaps we could make the actual `var`s -> `private var`s to force users to use the getters / setters (and provide an additional level of indirection separating interface from implementation)

rotationsymmetry replied:
Got it. That is a good point. I will make them private in my next push to the PR.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,57 @@
+    if (timeUnit != StreamingDecay.BATCHES && timeUnit != StreamingDecay.POINTS) {
+      throw new IllegalArgumentException("Invalid time unit for decay: " + timeUnit)
+    }
+    this.decayFactor = math.exp(math.log(0.5) / halfLife)
```
The formula for calculating `decayFactor` needs to be documented

rotationsymmetry replied:
will include this in the ScalaDoc. Thanks!

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearAlgorithm.scala:
```diff
@@ -84,7 +87,20 @@ abstract class StreamingLinearAlgorithm[
+        val discount = getDiscount(numNewDataPoints)
+
+        val updatedDataWeight = previousDataWeight * discount + numNewDataPoints
+        val lambda = numNewDataPoints / math.max(updatedDataWeight, 1e-16)
```
Assuming `previousDataWeight >= 0` and `discount >= 0`, `updatedDataWeight >= numNewDataPoints` so there's no need for protecting against div by zero here.

We should remove the check but keep a comment here in case readers of the code worry about it

feynmanliang replied:
I actually think it's more clear if `StreamingDecay` had a `getLambda`, otherwise you will need to document how `discount` is calculated in `StreamingDecay` as well as how `lambda` affects the weight updates in `StreamingLinearAlgorithm`

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionSuite.scala:
```diff
@@ -184,4 +184,72 @@ class StreamingLogisticRegressionSuite extends SparkFunSuite with TestSuiteBase
+    runStreams(ssc, numBatches, numBatches)
+
+    // with full memory, the final parameter estimates should be close to model A
+    assert(model.latestModel().weights(0) ~== 0.5 relTol 0.5)
```
A 50% tolerance is quite large, can the tests still regularly pass with a smaller tolerance?

rotationsymmetry replied:
In the test "parameter accuracy with full memory (decayFactor = 1)", all RDD effectively contribute equally to the final estimate. Since the last RDD is generated with a different model than the previous RDD, the final estimate has more variability than usual.

In the test "parameter accuracy with no memory (decayFactor = 0)", effectively all the previous RDD are ignored and the algorithm only has the samples from the last RDD to work with. Therefore the sample size provided to the algorithm is small and the estimate could fluctuate quite a bit. 

In summary, for a smaller tolerance, we basically need to reduce the variability of the final estimate. I have increased the number of records in each RDD in the stream along with small tweak to the parameters. Now we can pass these tests consistently with 10% tolerance. Will include these update in my next push to the PR.

feynmanliang replied:
Awesome, thanks for investigating! I just saw decayFactor=1 and I assumed that the weights would be ~ (1/numBatches = 1 / 20 = 0.05) off from the true value

feynmanliang commented:
@rotationsymmetry you also have a merge conflict, sorry :disappointed: do you mind resolving?

rotationsymmetry committed:
incorporating comments

Add ScalaDoc for public API.
Add ScalaDoc to decribe the forgetful algorithm in StreamingLinearAlgorithm.
Remove F-polymorphism in StreamingDecaySetter[T].
decayFactor and timeUnit in StreamingDecaySetter[T] are now private.
Remove division by zero in trainOn of StreamingLinearAlgorithm; provide comments to explains why.
Improve testing cases of StreamingLogisticRegressionSuite to have rel tol=0.1.

rotationsymmetry commented:
@feynmanliang Thank you very much for your review. 

I have incorporated your comments in commit a4ed2b0.

* Add ScalaDoc for public API.
* Add ScalaDoc to decribe the forgetful algorithm in StreamingLinearAlgorithm.
* Remove F-polymorphism in StreamingDecaySetter[T].
* decayFactor and timeUnit in StreamingDecaySetter[T] are now private.
* Remove division by zero in trainOn of StreamingLinearAlgorithm; provide comments to explains why.
* Improve testing cases of StreamingLogisticRegressionSuite to have rel tol=0.1.
* resolve merge conflict.

As for your comment of "having getLambda instead of getDiscount in StreamingDecay", I feel that the discount factor better conveys the mathematical idea of the algorithm. Lambda, on the other hand, is only a temporary value in the calculation. For example, in the [spark doc](http://spark.apache.org/docs/latest/mllib-clustering.html#streaming-k-means), the discount factor is employed to describe the algorithm. I have included similar description in the ScalaDoc for `StreamingLinearAlgorithm`.

Thanks again for your review. If you have any further comments, please let me know.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionWithSGD.scala:
```diff
@@ -32,6 +32,11 @@ import org.apache.spark.mllib.regression.StreamingLinearAlgorithm
  * of features must be constant. An initial weight
  * vector must be provided.
  *
+ * This class inherits the forgetful algorithm from StreamingLinearAlgorithm
```
"[[StreamingLinearAlgorithm]]" so API docs generate a link, ditto for L37

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+ * :: Experimental ::
+ * Supplies an interface for the discount value in
+ * the forgetful update rule in StreamingLinearAlgorithm.
+ * Actual implementation is provided in StreamingDecaySetter[T].
```
"[[StreamingDecaySetter]]"

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+ * Actual implementation is provided in StreamingDecaySetter[T].
+ */
+@Experimental
+trait StreamingDecay {
```
`private[mllib]` (we can always relax to `private[spark]` if we need this trait in `ml` but not vice versa)

feynmanliang replied:
Actually, do we plan on adding more `StreamingDecay` implementations soon? If not, I would suggest denormalizing and just having a single trait rather than this hierarchy

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+ * and half-life.
+ */
+@Experimental
+private[mllib] trait StreamingDecaySetter[T] extends Logging {
```
If `StreamingDecaySetter` implements `StreamingDecay` like L27 says, then this should be `trait StreamingDecaySetter[T] extends StreamingDecay with Logging`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+ */
+@Experimental
+private[mllib] trait StreamingDecaySetter[T] extends Logging {
+  self: T =>
```
keep self type annotation on L47

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+   * The decay factor is calculated such that, for data acquired at time t,
+   * its contribution by time t + halfLife will have dropped to 0.5.
+   * The unit of time can be specified either as batches or points;
+   * see StreamingDecay companion object.
```
I don't think we need to add reference to companion object since it doesn't add any information (plus, I want to make the companion package private)

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+ * Provides the String constants for allowed time unit in the forgetful algorithm.
+ */
+@Experimental
+object StreamingDecay {
```
Can this be `private[mllib]`?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+   * Set the half life and time unit ("batches" or "points") for the forgetful algorithm.
+   * The half life along with the time unit provides an alternative way to specify decay factor.
+   * The decay factor is calculated such that, for data acquired at time t,
+   * its contribution by time t + halfLife will have dropped to 0.5.
```
nit: "dropped to 0.5" -> "dropped by 0.5"

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,116 @@
+   * Each RDD in the DStream will be treated as 1 time unit.
+   *
+   */
+  final val BATCHES = "batches"
```
Looks like you're trying to get an enum-like thing here using strings? I would tend to favor case objects in this case, something like

```scala
object StreamingDecay {
  sealed trait TimeUnit
  case object BATCHES extends TimeUnit
  case object POINTS extends TimeUnit
}
```

and then use `StreamingDecay.TimeUnit` instead of `String` for the type

rotationsymmetry replied:
I am all for this approach because if offers much higher type safety and the IDE goodies. 

The reason I have the `String` implementation is to follow [StreamingKMeans](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.mllib.clustering.StreamingKMeans). 

Shall we consolidate StreamingKMeans to use case object? If so, I will open a JIRA for that.

feynmanliang replied:
I think that would be much better but it might be useful to ping whoever implemented the StreamingKMeans API first just to check if they did that for some particular reason that I am unaware of.

rotationsymmetry replied:
@freeman-lab : 

@feynmanliang and I are working on adding forgetful algorithm with decay factors to the streaming regressions. We would propose to implement case object for the time unit of half life in streaming algorithms. The case object approach will enhance type safety and readability. Currently `StreamingKMeans` uses `String` constant ("BATCHES", "POINTS") for the time unit.  Check if this proposal is OK with you? 

Proposed case objects:
``` scala
  sealed trait TimeUnit
  case object BATCHES extends TimeUnit
  case object POINTS extends TimeUnit
```

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearRegressionWithSGD.scala:
```diff
@@ -47,6 +52,7 @@ class StreamingLinearRegressionWithSGD private[mllib] (
     private var numIterations: Int,
     private var miniBatchFraction: Double)
   extends StreamingLinearAlgorithm[LinearRegressionModel, LinearRegressionWithSGD]
+  with StreamingDecaySetter[StreamingLinearRegressionWithSGD]
```
OK, I think I get why you have both `StreamingDecay` and `StreamingDecaySetter`; it's because `getDiscount` is also needed in `trainOn` but you also need to get the concrete type `StreamingLinearRegressionWithSGD` for returning from the setters (e.g. `setHalfLife`), right?

I'm still for having a single trait over trait inheritance, especially if the reason we have this inheritance is to get the concrete type and we won't be adding other concrete `StreamingDecay` implementations soon.

We could just have a single `private[mllib] trait StreamingDecay` and have `StreamingLinearAlgorithm extends StreamingDecay[A]`; the concrete type should be available in `A` and return properly. Let me know if that works...

rotationsymmetry replied:
Thanks for the suggestions. I agree we can consolidate `StreamingDecay` and `StreamingDecaySetter` into one class. 

Regarding the implementation. If we do the following
``` scala
trait StreamingDecay[T] {
self: T =>
  def setX: T = this
}
class StreamingLinearAlgorithm extends StreamingDecay[StreamingLinearAlgorithm]
class StreamingLinearRegressionWithSGD extends StreamingLinearAlgorithm
val s = new StreamingLinearRegressionWithSGD()
```
Then the return type of `s.setX` will be `StreamingLinearAlgorithm` since this is what the generic `T` refers to.

So I propose we override the `setX` method to get the correct type. 

``` scala
trait StreamingDecay {
  def setX: this.type = this
}
class StreamingLinearAlgorithm extends StreamingDecay
class StreamingLinearRegressionWithSGD extends StreamingLinearAlgorithm {
  override def setX: this.type = {
    super.setX
    this
  } 
}
val s = new StreamingLinearRegressionWithSGD()
```

I have run this proposed implementation and the code is working. Shall we proceed?

feynmanliang replied:
Why can't we do

```scala
abstract class StreamingLinearAlgorithm[
    M <: GeneralizedLinearModel,
    A <: GeneralizedLinearAlgorithm[M]] extends Logging with StreamingDecay[A] {
```
since `A` should have the concrete subtype

rotationsymmetry replied:
Thanks for the suggestions. I think `A` captures the class of non-streaming algorithm. For example, 
``` scala
class StreamingLinearRegressionWithSGD 
  extends StreamingLinearAlgorithm[LinearRegressionModel, LinearRegressionWithSGD]
```
The concrete subtype that `A` capture is `LinearRegressionWithSGD`. However, the fluent setter needs to return `StreamingLinearRegressionWithSGD`, though.

What do you think?

feynmanliang replied:
Aah you're right... In that case, I think what you have is a fine solution.

rotationsymmetry committed:
incorporating further comments

Refactor StreamingDecay
Use case object for TimeUnit
Clean up ScalaDoc

rotationsymmetry commented:
@feynmanliang I have make another push to the PR:

Refactor StreamingDecay
Use case object for TimeUnit
Clean up ScalaDoc

Thank again for your review.

rotationsymmetry committed:
Add tests for half-life and TimeUnit.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionWithSGD.scala:
```diff
@@ -101,4 +107,14 @@ class StreamingLogisticRegressionWithSGD private[mllib] (
     this
   }
+
+  override def setDecayFactor(decayFactor: Double): this.type = {
```
This boilerplate is duplicated in streaming linear regression. I am guessing you do this to get the concrete subclass (correct me if I'm wrong), but you actually don't need to do this since the `this.type` in `trait StreamingDecay` takes care of this. A simple REPL example:

```scala
scala> trait Superclass { def test: this.type }
defined trait Superclass

scala> class Subclass extends Superclass { def test = this }
defined class Subclass

scala> (new Subclass()).test
res0: Subclass = Subclass@1cb4ab3e
```

rotationsymmetry replied:
fixed.

feynmanliang replied:
Oh, I meant that you could remove these setters entirely

```scala
scala> trait Superclass { def test: this.type = this }
defined trait Superclass

scala> class Subclass extends Superclass
defined class Subclass


scala> (new Subclass).test
res1: Subclass = Subclass@b364520
```

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,99 @@
+    this.decayFactor = decayFactor
+    this
+  }
+
```
Extra newline

rotationsymmetry replied:
fixed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,99 @@
+   *
+   * @param decayFactor the decay factor
+   */
+  def setDecayFactor(decayFactor: Double): this.type = {
```
Since these are part of public API, please add `@Since` annotations

feynmanliang replied:
We should also document default values for `decayFactor` and `timeUnit`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * Set the decay factor for the forgetful algorithms.
+   * The decay factor should be between 0 and 1, inclusive.
+   * decayFactor = 0: only the data from the most recent RDD will be used.
```
This is only true if `TimeUnit = BATCHES`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,99 @@
+   * @return Discount factor
+   */
+  private[mllib] def getDiscount(numNewDataPoints: Long): Double = timeUnit match {
+    case BATCHES => decayFactor
```
Since `timeUnit` affects the behavior `decayFactor` as well, what do you think about either:
1. adding a `timeUnit` param to `decayFactor`
2. removing the `timeUnit` param from `setHalfLife` and introducing a `setTimeUnit`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearRegressionWithSGD.scala:
```diff
@@ -107,4 +113,16 @@ class StreamingLinearRegressionWithSGD private[mllib] (
     this
   }
+
+  override def setDecayFactor(decayFactor: Double): this.type = {
```
Ditto on duplication

rotationsymmetry replied:
fixed.

feynmanliang replied:
Can remove these setters as well since trait will return concrete subtype

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/regression/StreamingLinearRegressionSuite.scala:
```diff
@@ -17,6 +17,8 @@
 
+import org.apache.spark.mllib.regression.StreamingDecay.{BATCHES, POINTS}
+
 import scala.collection.mutable.ArrayBuffer
```
organize imports

rotationsymmetry replied:
fixed.

feynmanliang commented on mllib/src/test/scala/org/apache/spark/mllib/regression/StreamingLinearRegressionSuite.scala:
```diff
@@ -194,4 +196,204 @@ class StreamingLinearRegressionSuite extends SparkFunSuite with TestSuiteBase {
+    runStreams(ssc, numBatches, numBatches)
+
+    // with no memory, the final parameter estimates should be close to model B
+    assertEqual(model.latestModel().intercept, 0.0, 1.0)
```
Why not `~==`?

rotationsymmetry replied:
Sure.

feynmanliang commented:
Made another pass

rotationsymmetry committed:
Refactor: timeUnit has its own setter.

Add @Since.
Clean up ScalaDoc.

rotationsymmetry commented:
@feynmanliang Thank you for your comments. I have revised the PR, including
* Refactor: timeUnit has its own setter.
* Add @Since.
* Clean up ScalaDoc.

As I am rewriting the ScalaDoc, it appears that the algorithm can be more easily described and understood if we use the parameter RetentionFactor (aka, 1- decayFactor) instead of decayFactor. What do you think?

feynmanliang commented:
Streaming KMeans uses `decayFactor` and I think it's important we maintain consistency

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,117 @@
+ */
+@Experimental
+private[mllib] trait StreamingDecay extends Logging{
+  private[this] var decayFactor: Double = 0
```
just `private` is fine

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,117 @@
+  private[mllib] sealed trait TimeUnit
+  /**
+   * Each RDD in the DStream will be treated as 1 time unit.
+   *
```
extra newline

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,117 @@
+  case object BATCHES extends TimeUnit
+  /**
+   * Each data point will be treated as 1 time unit.
+   *
```
Extra newline

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingDecay.scala:
```diff
@@ -0,0 +1,117 @@
+ * forgetful update rule in [[StreamingLinearAlgorithm]];
+ * The degree of forgetfulness can be specified by the decay factor
+ * or the half life.
+ *
```
Extra newline

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearAlgorithm.scala:
```diff
@@ -59,11 +76,14 @@ import org.apache.spark.streaming.dstream.DStream
   /** The model to be updated and used for prediction. */
   protected var model: Option[M]
 
+  protected var previousDataWeight: Double = 0
```
nit: Add a comment for this to be consistent with the other `var` declarations

feynmanliang commented:
LGTM after these changes and pending tests

CC @mengxr @freeman-lab

rotationsymmetry committed:
remove duplicate setters.

clean up new lines and comments.

rotationsymmetry commented:
@feynmanliang Much appreciated. I have update the PR for your comments.

mengxr commented:
add to whitelist

mengxr commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44017 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44017/consoleFull)** for PR 8022 at commit [`9ba83cb`](https://github.com/apache/spark/commit/9ba83cbc6a45a772558211680e6c1f1933f6d8f6).

SparkQA commented:
**[Test build #44017 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44017/consoleFull)** for PR 8022 at commit [`9ba83cb`](https://github.com/apache/spark/commit/9ba83cbc6a45a772558211680e6c1f1933f6d8f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44017/
Test PASSed.

mengxr commented:
@rotationsymmetry Could you provide a simple unit test in Java to show Java compatibility?

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearAlgorithm.scala:
```diff
@@ -91,7 +112,22 @@ abstract class StreamingLinearAlgorithm[
+        val lambda = numNewDataPoints / updatedDataWeight
+
+        BLAS.scal(lambda, newModel.weights)
+        BLAS.axpy(1-lambda, model.get.weights, newModel.weights)
```
Do we have some references about this merging scheme? I assume that this works for many cases, but there is no guarantee in theory.

Commit:
Improve Java API compatibility.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45336 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45336/consoleFull)** for PR 8022 at commit [`0072400`](https://github.com/apache/spark/commit/0072400825c0c4a871467a92799d12752fe96402).

SparkQA commented:
**[Test build #45336 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45336/consoleFull)** for PR 8022 at commit [`0072400`](https://github.com/apache/spark/commit/0072400825c0c4a871467a92799d12752fe96402).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented:
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one.
